### PR TITLE
Correctly format typed self in function arguments

### PR DESCRIPTION
**Pull request by marcusklaas**
This closes https://github.com/nrc/rustfmt/issues/153.

### CODE REVIEW

File: src/items.rs [hunk 1]
```diff
@@ -790,6 +759,37 @@ impl<'a> FmtVisitor<'a> {
     }
 }
 
+fn rewrite_explicit_self(explicit_self: &ast::ExplicitSelf, args: &[ast::Arg]) -> String {
```
Comment (nrc) [inline src/items.rs]: Better to return `Option<String>` than giving a special meaning to the empty string (also saves an allocation).

Comment (marcusklaas) [inline src/items.rs]: I considered this, but reserved the `None` case for if we want to implement `Rewrite` for it later. I don't think `String::new` allocates, by the way.

File: src/items.rs [hunk 2]
```diff
@@ -232,42 +233,10 @@ impl<'a> FmtVisitor<'a> {
...
-                    };
-
-                    arg_item_strs[0] = format!("{}self", mut_str);
-                    min_args = 2;
-                }
-                _ => {}
+            let self_str = rewrite_explicit_self(explicit_self, args);
+            if self_str.len() > 0 {
+                arg_item_strs[0] = self_str;
+                min_args = 2;
```
Comment (nrc) [inline src/items.rs]: If I understand correctly, the only change here is that we now change min_args to 2 in the `SelfExplicit` case. Was that the bug before? I expected to see some change with formatting a missing span or something. Could you explain how the bug happened please?

Comment (nrc) [inline src/items.rs]: Also, and more of a nit, I think it is better to initialise min_args here, rather than having a mutable variable.

**Comment by nrc** [comment]
@marcusklaas thanks for all the patches! You've done a tonne of good work recently :-) Heads up that I'll be travelling next week, so might be a bit slower with reviews and merging.

File: src/items.rs (hunk 1 above)
Comment (nrc) [inline src/items.rs]: I guess maybe return a custom enum then? Alternatively we could always return Option<Option<String>> if we implement Rewrite (which would require making Rewrite more generic I guess). If we have to treat the result specially if its an empty string, then it really ought to be explicit by using an enum of some kind.

File: src/items.rs (hunk 2 above)
Comment (marcusklaas) [inline src/items.rs]: It's a bad span: https://github.com/rust-lang/rust/issues/27522. The item is skipped when `min_args == 2`.

**Comment by marcusklaas** [comment]
Updated!

**Comment by marcusklaas** [comment]
Wait, may be there is no need to skip it once we fix the span... Why do we need `min_args = 2` for the other cases?

**Comment by marcusklaas** [comment]
Should we consider merging this? It's not ideal, but at least fixes the underlying bug. To do it the proper way would be to address https://github.com/rust-lang/rust/issues/27522, which isn't trivial.

Until that happens, the comment for the self type is dropped. We could open a separate issue for that.

**Comment by nrc** [comment]
Yes, we should! (Sorry it took a while to get to).

@marcusklaas could you add a fixme comment referencing #27522 please? Then you can push this.

**Pull request merged**